core/src/local/persistent_state.rs

```diff
@@ -0,0 +1,891 @@
+    fn in_domain_fn(key: &[u8]) -> bool {
+        key != META_KEY
+    }
+
+    // A key is built up of five components:
+    //  * `index_id`
+    //     Uniquely identifies the index in self.indices.
+    //  * `row_size`
+    //      The byte size of `row` when serialized with binode.
+    //  * `row`
+    //      The actual index values.
+    //  * `epoch`
+    //      Incremented on each PersistentState initialization.
+    //  * `seq`
+    //      Sequence number since last epoch.
+    fn serialize_key(&self, index_id: IndexID, row: &Vec<&DataType>, seq: IndexSeq) -> Vec<u8> {
+        let size: u64 = bincode::serialized_size(&row).unwrap();
+        bincode::serialize(&(index_id, size, row, self.epoch, seq)).unwrap()
+    }
+
+    // Used with DB::prefix_iterator to go through all the rows for a given key.
+    fn serialize_prefix(index: IndexID, key: &Vec<&DataType>) -> Vec<u8> {
+        let size: u64 = bincode::serialized_size(&key).unwrap();
+        bincode::serialize(&(index, size, key)).unwrap()
+    }
+
+    // Filters out secondary indices to return an iterator for the actual key-value pairs.
+    fn all_rows(&self) -> impl Iterator<Item = (Box<[u8]>, Box<[u8]>)> {
+        self.db
+            .as_ref()
+            .unwrap()
+            .full_iterator(rocksdb::IteratorMode::Start)
+            .filter(|&(ref key, _)| {
+                // Filter out non-pk indices:
+                let i: IndexID = bincode::deserialize(&key).unwrap();
+                i == 0
+            })
+    }
+
+    // Puts by primary key first, then retrieves the existing value for each index and appends the
+    // newly created primary key value.
+    // TODO(ekmartin): This will put exactly the values that are given, and can only be retrieved
+    // with exactly those values. I think the regular state implementation supports inserting
+    // something like an Int and retrieving with a BigInt.
+    fn insert(&mut self, batch: &mut WriteBatch, r: Vec<DataType>) {
+        for index in self.indices.iter_mut() {
+            index.seq += 1;
+        }
+
+        let (secondary_indices, serialized_pk) = if let Some(ref pk_cols) = self.primary_key {
+            let pk = pk_cols.iter().map(|i| &r[*i]).collect::<Vec<_>>();
+            (&self.indices[..], Self::serialize_prefix(0, &pk))
+        } else {
+            // For bases without primary keys we store the actual row values keyed by the index
+            // that was added first. This means that we can't consider the keys unique though, so
+            // we'll append a sequence number.
+            let pk_index = &self.indices[0];
+            let pk = pk_index.columns.iter().map(|i| &r[*i]).collect::<Vec<_>>();
+            (&self.indices[1..], self.serialize_key(0, &pk, pk_index.seq))
+        };
+
+        // First insert the actual value for our primary index:
+        let pk_value = bincode::serialize(&r).unwrap();
+        batch.put(&serialized_pk, &pk_value).unwrap();
```

Comment:
I'm a little surprised that the `rocksdb` don't have a way of sending a value without storing it in an intermediate `&[u8]` buffer... Oh well.